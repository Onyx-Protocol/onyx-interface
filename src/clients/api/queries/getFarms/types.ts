import BigNumber from 'bignumber.js';
import { Multicall } from 'ethereum-multicall';
import { EthChainId, Token } from 'types';

export interface Address {
  [EthChainId.MAINNET]: string;
  [EthChainId.TESTNET]: string;
}

export interface FarmConfig {
  pid: number;
  lpSymbol: Uppercase<string>;
  lpAddresses: Address;
  token: Token;
  quoteToken: Token;
  multiplier?: string;
}

export interface Farm extends FarmConfig {
  tokenAmount?: BigNumber;
  quoteTokenAmount?: BigNumber;
  lpTotalInQuoteToken?: BigNumber;
  tokenPriceVsQuote?: BigNumber;
  poolWeight?: BigNumber;
  lpTokenBalanceMC?: BigNumber;
  tokenPerSecond?: BigNumber;
  farmApr?: number;
  userData?: {
    allowance: BigNumber;
    tokenBalance: BigNumber;
    stakedBalance: BigNumber;
    earnings: BigNumber;
  };
}

export interface GetFarmsInput {
  multicall: Multicall;
  accountAddress: string;
}

export type GetFarmsOutput = {
  farms: Farm[];
};

export const getAddress = (address: Address): string => {
  const chainId: EthChainId = process.env.REACT_APP_CHAIN_ID
    ? parseInt(process.env.REACT_APP_CHAIN_ID, 10)
    : EthChainId.MAINNET;
  return address[chainId];
};
