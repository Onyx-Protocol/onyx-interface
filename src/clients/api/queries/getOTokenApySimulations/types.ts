import BigNumber from 'bignumber.js';
import { Multicall } from 'ethereum-multicall';

export interface GetOTokenInterestRatesInput {
  multicall: Multicall;
  reserveFactorMantissa: BigNumber;
  interestRateModelContractAddress: string;
}

export interface OTokenApySnapshot {
  utilizationRate: number;
  borrowApyPercentage: number;
  supplyApyPercentage: number;
}

export type GetOTokenApySimulationsOutput = {
  apySimulations: OTokenApySnapshot[];
};
