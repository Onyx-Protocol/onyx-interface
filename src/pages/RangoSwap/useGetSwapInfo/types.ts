import { Trade as PSTrade } from '@uniswap/sdk';
import { Swap, SwapDirection, Token } from 'types';

export type SwapError = 'INSUFFICIENT_LIQUIDITY' | 'WRAPPING_UNWRAPPING_UNSUPPORTED';

export interface UseGetSwapInfoInput {
  fromToken: Token;
  toToken: Token;
  direction: SwapDirection;
  fromTokenAmountTokens?: string;
  toTokenAmountTokens?: string;
}

export interface UseGetSwapInfoOutput {
  swap: Swap | undefined;
  error: SwapError | undefined;
}

export interface FormatToSwapInput {
  input: UseGetSwapInfoInput;
  trade: PSTrade;
}

export type FormatToSwapOutput = Swap;
