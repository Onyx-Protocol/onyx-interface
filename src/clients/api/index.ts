export { default as queryClient } from './queryClient';

// Mutations
export { default as enterMarkets } from './mutations/enterMarkets';
export * from './mutations/enterMarkets';
export { default as useEnterMarkets } from './mutations/enterMarkets/useEnterMarkets';

export { default as exitMarket } from './mutations/exitMarket';
export * from './mutations/exitMarket';
export { default as useExitMarket } from './mutations/exitMarket/useExitMarket';

export { default as approveToken } from './mutations/approveToken';
export * from './mutations/approveToken';
export { default as useApproveToken } from './mutations/approveToken/useApproveToken';

export { default as approveNftForAll } from './mutations/approveNftForAll';
export * from './mutations/approveNftForAll';
export { default as useApproveNftForAll } from './mutations/approveNftForAll/useApproveNftForAll';

export { default as supplyNonEth } from './mutations/supplyNonEth';
export * from './mutations/supplyNonEth';
export { default as useSupplyNonEth } from './mutations/supplyNonEth/useSupplyNonEth';
export * from './mutations/supplyNonEth/useSupplyNonEth';

export { default as supplyEth } from './mutations/supplyEth';
export * from './mutations/supplyEth';
export { default as useSupplyEth } from './mutations/supplyEth/useSupplyEth';
export * from './mutations/supplyEth/useSupplyEth';

export { default as supplyNFT } from './mutations/supplyNFT';
export * from './mutations/supplyNFT';
export { default as useSupplyNFT } from './mutations/supplyNFT/useSupplyNFT';
export * from './mutations/supplyNFT/useSupplyNFT';

export { default as redeem } from './mutations/redeem';
export * from './mutations/redeem';
export { default as useRedeem } from './mutations/redeem/useRedeem';

export { default as redeemNFT } from './mutations/redeemNFT';
export * from './mutations/redeemNFT';
export { default as useRedeemNFT } from './mutations/redeemNFT/useRedeemNFT';

export { default as repayNonEthOToken } from './mutations/repayNonEthOToken';
export * from './mutations/repayNonEthOToken';
export { default as useRepayNonEthOToken } from './mutations/repayNonEthOToken/useRepayNonEthOToken';

export { default as repayEth } from './mutations/repayEth';
export * from './mutations/repayEth';
export { default as useRepayEth } from './mutations/repayEth/useRepayEth';

export { default as redeemUnderlying } from './mutations/redeemUnderlying';
export * from './mutations/redeemUnderlying';
export { default as useRedeemUnderlying } from './mutations/redeemUnderlying/useRedeemUnderlying';

export { default as claimXcnReward } from './mutations/claimXcnReward';
export * from './mutations/claimXcnReward';
export { default as useClaimXcnReward } from './mutations/claimXcnReward/useClaimXcnReward';

export { default as borrowOToken } from './mutations/borrowOToken';
export * from './mutations/borrowOToken';
export { default as useBorrowOToken } from './mutations/borrowOToken/useBorrowOToken';

export { default as useRepayOToken } from './mutations/useRepayOToken';

export { default as createProposal } from './mutations/createProposal';
export * from './mutations/createProposal';
export { default as useCreateProposal } from './mutations/createProposal/useCreateProposal';

export { default as cancelProposal } from './mutations/cancelProposal';
export * from './mutations/cancelProposal';
export { default as useCancelProposal } from './mutations/cancelProposal/useCancelProposal';

export { default as executeProposal } from './mutations/executeProposal';
export * from './mutations/executeProposal';
export { default as useExecuteProposal } from './mutations/executeProposal/useExecuteProposal';

export { default as queueProposal } from './mutations/queueProposal';
export * from './mutations/queueProposal';
export { default as useQueueProposal } from './mutations/queueProposal/useQueueProposal';

export { default as castVote } from './mutations/vote/castVote';
export * from './mutations/vote/castVote';
export { default as useCastVote } from './mutations/vote/useCastVote';

export { default as registerProxy } from './mutations/registerProxy';
export * from './mutations/registerProxy';
export { default as useRegisterProxy } from './mutations/registerProxy/useRegisterProxy';

export { default as depositPunk } from './mutations/depositPunk';
export * from './mutations/depositPunk';
export { default as useDepositPunk } from './mutations/depositPunk/useDepositPunk';

export { default as mintWPunks } from './mutations/mintWPunks';
export * from './mutations/mintWPunks';
export { default as useMintWPunks } from './mutations/mintWPunks/useMintWPunks';

export { default as burnWPunks } from './mutations/burnWPunks';
export * from './mutations/burnWPunks';
export { default as useBurnWPunks } from './mutations/burnWPunks/useBurnWPunks';

export { default as swapTokens } from './mutations/swapTokens';
export * from './mutations/swapTokens';
export { default as useSwapTokens } from './mutations/swapTokens/useSwapTokens';

// Queries
export { default as getAssetsInAccount } from './queries/getAssetsInAccount';
export * from './queries/getAssetsInAccount';
export { default as useGetAssetsInAccount } from './queries/getAssetsInAccount/useGetAssetsInAccount';

export { default as getHypotheticalAccountLiquidity } from './queries/getHypotheticalAccountLiquidity';
export * from './queries/getHypotheticalAccountLiquidity';

export { default as getMarkets } from './queries/getMarkets';
export * from './queries/getMarkets';
export { default as useGetMarkets } from './queries/getMarkets/useGetMarkets';

export { default as getOTokenBalancesAll } from './queries/getOTokenBalancesAll';
export * from './queries/getOTokenBalancesAll';
export { default as useGetOTokenBalancesAll } from './queries/getOTokenBalancesAll/useGetOTokenBalancesAll';

export { default as getOTokenBalanceOf } from './queries/getOTokenBalanceOf';
export * from './queries/getOTokenBalanceOf';
export { default as useGetOTokenBalanceOf } from './queries/getOTokenBalanceOf/useGetOTokenBalanceOf';

export { default as getXcnReward } from './queries/getXcnReward';
export * from './queries/getXcnReward';
export { default as useGetXcnReward } from './queries/getXcnReward/useGetXcnReward';

export { default as getAllowance } from './queries/getAllowance';
export * from './queries/getAllowance';
export { default as useGetAllowance } from './queries/getAllowance/useGetAllowance';

export { default as getIsApprovedForAll } from './queries/getIsApprovedForAll';
export * from './queries/getIsApprovedForAll';
export { default as useGetIsApprovedForAll } from './queries/getIsApprovedForAll/useGetIsApprovedForAll';

export { default as getBalanceOf } from './queries/getBalanceOf';
export * from './queries/getBalanceOf';
export { default as useGetBalanceOf } from './queries/getBalanceOf/useGetBalanceOf';

export { default as getTokenBalances } from './queries/getTokenBalances';
export * from './queries/getTokenBalances';
export { default as useGetTokenBalances } from './queries/getTokenBalances/useGetTokenBalances';

export { default as useGetUserMarketInfo } from './queries/useGetUserMarketInfo';

export { default as useGetTreasuryTotals } from './queries/useGetTreasuryTotals';

export { default as getMarketHistory } from './queries/getMarketHistory';
export * from './queries/getMarketHistory';
export { default as useGetMarketHistory } from './queries/getMarketHistory/useGetMarketHistory';

export { default as getOTokenCash } from './queries/getOTokenCash';
export * from './queries/getOTokenCash';
export { default as useGetOTokenCash } from './queries/getOTokenCash/useGetOTokenCash';

export { default as getOTokenInterestRateModel } from './queries/getOTokenInterestRateModel';
export * from './queries/getOTokenInterestRateModel';
export { default as useGetOTokenInterestRateModel } from './queries/getOTokenInterestRateModel/useGetOTokenInterestRateModel';

export { default as getOTokenApySimulations } from './queries/getOTokenApySimulations';
export * from './queries/getOTokenApySimulations';
export { default as useGetOTokenApySimulations } from './queries/getOTokenApySimulations/useGetOTokenApySimulations';

export { default as getCurrentVotes } from './queries/getCurrentVotes';
export * from './queries/getCurrentVotes';
export { default as useGetCurrentVotes } from './queries/getCurrentVotes/useGetCurrentVotes';

export { default as getStakingInfos } from './queries/getStakingInfos';
export * from './queries/getStakingInfos';
export { default as useGetStakingInfos } from './queries/getStakingInfos/useGetStakingInfos';

export { default as getStakingApy } from './queries/getStakingApy';
export * from './queries/getStakingApy';
export { default as useGetStakingApy } from './queries/getStakingApy/useGetStakingApy';

export { default as getTransactions } from './queries/getTransactions';
export * from './queries/getTransactions';
export { default as useGetTransactions } from './queries/getTransactions/useGetTransactions';

export { default as getDailyXcn } from './queries/getDailyXcn';
export * from './queries/getDailyXcn';
export { default as useGetDailyXcn } from './queries/getDailyXcn/useGetDailyXcn';

export { default as getProposals } from './queries/getProposals';
export * from './queries/getProposals';
export { default as useGetProposals } from './queries/getProposals/useGetProposals';

export * from './queries/getStakeHistories';
export { default as useGetStakeHistories } from './queries/getStakeHistories/useGetStakeHistories';

export { default as getProposal } from './queries/getProposals/getProposal';
export * from './queries/getProposals/getProposal';
export { default as useGetProposal } from './queries/getProposals/useGetProposal';

export { default as getVoteReceipt } from './queries/getVoteReceipt';
export * from './queries/getVoteReceipt';
export { default as useGetVoteReceipt } from './queries/getVoteReceipt/useGetVoteReceipt';

export { default as getVoters } from './queries/getVoters';
export * from './queries/getVoters';
export { default as useGetVoters } from './queries/getVoters/useGetVoters';

export { default as getVoterDetails } from './queries/getVoterDetails';
export * from './queries/getVoterDetails';
export { default as useGetVoterDetails } from './queries/getVoterDetails/useGetVoterDetails';

export { default as getVoterHistory } from './queries/getVoterHistory';
export * from './queries/getVoterHistory';
export { default as useGetVoterHistory } from './queries/getVoterHistory/useGetVoterHistory';

export { default as getVoterAccounts } from './queries/getVoterAccounts';
export * from './queries/getVoterAccounts';
export { default as useGetVoterAccounts } from './queries/getVoterAccounts/useGetVoterAccounts';

export { default as getProposalThreshold } from './queries/getProposalThreshold';
export * from './queries/getProposalThreshold';
export { default as useGetProposalThreshold } from './queries/getProposalThreshold/useGetProposalThreshold';

export { default as getProposalState } from './queries/getProposalState';
export * from './queries/getProposalState';
export { default as useGetProposalState } from './queries/getProposalState/useGetProposalState';

export { default as getLatestProposalIdByProposer } from './queries/getLatestProposalIdByProposer';
export * from './queries/getLatestProposalIdByProposer';
export { default as useGetLatestProposalIdByProposer } from './queries/getLatestProposalIdByProposer/useGetLatestProposalIdByProposer';

export { default as getBlockNumber } from './queries/getBlockNumber';
export * from './queries/getBlockNumber';
export { default as useGetBlockNumber } from './queries/getBlockNumber/useGetBlockNumber';

export { default as getProposalEta } from './queries/getProposalEta';
export * from './queries/getProposalEta';
export { default as useGetProposalEta } from './queries/getProposalEta/useGetProposalEta';

export { default as getProxies } from './queries/getProxies';
export * from './queries/getProxies';
export { default as useGetProxies } from './queries/getProxies/useGetProxies';

export { default as getOwnedPunkIds } from './queries/getOwnedPunkIds';
export * from './queries/getOwnedPunkIds';
export { default as useGetOwnedPunkIds } from './queries/getOwnedPunkIds/useGetOwnedPunkIds';

export { default as getOwnedWPunksIds } from './queries/getOwnedWPunksIds';
export * from './queries/getOwnedWPunksIds';
export { default as useGetOwnedWPunksIds } from './queries/getOwnedWPunksIds/useGetOwnedWPunksIds';

export { default as liquidateWithSingleRepay } from './mutations/liquidateWithSingleRepay';
export * from './mutations/liquidateWithSingleRepay';
export { default as useLiquidateWithSingleRepay } from './mutations/liquidateWithSingleRepay/useLiquidateWithSingleRepay';

export { default as liquidateBorrow } from './mutations/liquidateBorrow';
export * from './mutations/liquidateBorrow';
export { default as useLiquidateBorrow } from './mutations/liquidateBorrow/useLiquidateBorrow';

export { default as stakeXcn } from './mutations/stakeXcn';
export * from './mutations/stakeXcn';
export { default as useStakeXcn } from './mutations/stakeXcn/useStakeXcn';

export { default as withdrawXcn } from './mutations/withdrawXcn';
export * from './mutations/withdrawXcn';
export { default as useWithdrawXcn } from './mutations/withdrawXcn/useWithdrawXcn';

export { default as claimXcn } from './mutations/claimXcn';
export * from './mutations/claimXcn';
export { default as useClaimXcn } from './mutations/claimXcn/useClaimXcn';

export { default as getUniSwapPairs } from './queries/getUniSwapPairs';
export * from './queries/getUniSwapPairs';
export { default as useGetUniSwapPairs } from './queries/getUniSwapPairs/useGetUniSwapPairs';
