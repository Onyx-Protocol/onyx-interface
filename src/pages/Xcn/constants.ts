import BigNumber from 'bignumber.js';

import { TOKENS } from 'constants/tokens';

export const MINTED_XCN = '23700000';
export const MINTED_XCN_WEI = new BigNumber(MINTED_XCN).times(
  new BigNumber(10).pow(TOKENS.xcn.decimals),
);
