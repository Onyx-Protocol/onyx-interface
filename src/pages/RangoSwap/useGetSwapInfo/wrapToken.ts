import { Token } from 'types';

import { UNISWAP_TOKENS } from 'constants/tokens';

// UniSwap only trades with wrapped tokens, so ETH is replaced with wETH
const wrapToken = (token: Token) => (token.isNative ? UNISWAP_TOKENS.weth : token);

export default wrapToken;
