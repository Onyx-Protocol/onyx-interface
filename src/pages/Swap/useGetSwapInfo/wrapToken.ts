import { Token } from 'types';

import { UNISWAP_TOKENS } from 'constants/tokens';

// PancakeSwap only trades with wrapped tokens, so BNB is replaced with wBNB
const wrapToken = (token: Token) => (token.isNative ? UNISWAP_TOKENS.weth : token);

export default wrapToken;
