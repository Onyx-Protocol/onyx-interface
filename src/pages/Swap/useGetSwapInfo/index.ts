import {
  ChainId,
  Token as PSToken,
  TokenAmount as PSTokenAmount,
  Trade as PSTrade,
} from '@uniswap/sdk';
import BigNumber from 'bignumber.js';
import config from 'config';
import { useMemo } from 'react';
import { EthChainId } from 'types';
import { convertTokensToWei } from 'utilities';
import { toChecksumAddress } from 'web3-utils';

import { useGetUniSwapPairs } from 'clients/api';

import formatToSwap from './formatToSwap';
import { SwapError, UseGetSwapInfoInput, UseGetSwapInfoOutput } from './types';
import useGetTokenCombinations from './useGetTokenCombinations';
import wrapToken from './wrapToken';

export * from './types';

const useGetSwapInfo = (input: UseGetSwapInfoInput): UseGetSwapInfoOutput => {
  // Determine all possible token combinations based on input and base trade
  // tokens
  const tokenCombinations = useGetTokenCombinations({
    fromToken: input.fromToken,
    toToken: input.toToken,
  });

  // Fetch pair data
  const { data: getUniSwapPairsData } = useGetUniSwapPairs({ tokenCombinations });

  // Find the best trade based on pairs
  return useMemo(() => {
    let trade: PSTrade | undefined;
    let error: SwapError | undefined;

    const wrappedFromToken = wrapToken(input.fromToken);
    const wrappedToToken = wrapToken(input.toToken);

    // Return no trade if user is trying to wrap or unwrap ETH/wETH
    if (wrappedFromToken.address === wrappedToToken.address) {
      return {
        swap: undefined,
        error: 'WRAPPING_UNWRAPPING_UNSUPPORTED',
      };
    }

    // Handle "exactAmountIn" direction (sell an exact amount of fromTokens for
    // as many toTokens as possible)
    if (
      getUniSwapPairsData?.pairs &&
      getUniSwapPairsData?.pairs.length > 0 &&
      input.direction === 'exactAmountIn' &&
      !!input.fromTokenAmountTokens
    ) {
      const fromTokenAmountWei = convertTokensToWei({
        value: new BigNumber(input.fromTokenAmountTokens).dp(wrappedFromToken.decimals),
        token: wrappedFromToken,
      });

      const currencyAmountIn = new PSTokenAmount(
        new PSToken(
          config.chainId === EthChainId.MAINNET ? ChainId.MAINNET : ChainId.GÖRLI,
          toChecksumAddress(wrappedFromToken.address),
          wrappedFromToken.decimals,
          wrappedFromToken.symbol,
        ),
        fromTokenAmountWei.dp(wrappedFromToken.decimals).toFixed(),
      );

      const currencyOut = new PSToken(
        config.chainId === EthChainId.MAINNET ? ChainId.MAINNET : ChainId.GÖRLI,
        toChecksumAddress(wrappedToToken.address),
        wrappedToToken.decimals,
        wrappedToToken.symbol,
      );

      // Find best trade
      [trade] = PSTrade.bestTradeExactIn(
        getUniSwapPairsData?.pairs,
        currencyAmountIn,
        currencyOut,
        {
          maxHops: 3,
          maxNumResults: 1,
        },
      );

      error = trade ? undefined : 'INSUFFICIENT_LIQUIDITY';
    }

    // Handle "exactAmountOut" direction (sell as few fromTokens as possible for
    // a fixed amount of toTokens)
    if (
      getUniSwapPairsData?.pairs &&
      getUniSwapPairsData?.pairs.length > 0 &&
      input.direction === 'exactAmountOut' &&
      !!input.toTokenAmountTokens
    ) {
      const currencyIn = new PSToken(
        config.chainId === EthChainId.MAINNET ? ChainId.MAINNET : ChainId.GÖRLI,
        toChecksumAddress(wrappedFromToken.address),
        wrappedFromToken.decimals,
        wrappedFromToken.symbol,
      );

      const toTokenAmountWei = convertTokensToWei({
        value: new BigNumber(input.toTokenAmountTokens).dp(wrappedToToken.decimals),
        token: wrappedToToken,
      });

      const currencyAmountOut = new PSTokenAmount(
        new PSToken(
          config.chainId === EthChainId.MAINNET ? ChainId.MAINNET : ChainId.GÖRLI,
          toChecksumAddress(wrappedToToken.address),
          wrappedToToken.decimals,
          wrappedToToken.symbol,
        ),
        toTokenAmountWei.toFixed(),
      );

      // Find best trade
      [trade] = PSTrade.bestTradeExactOut(
        getUniSwapPairsData?.pairs,
        currencyIn,
        currencyAmountOut,
        {
          maxHops: 3,
          maxNumResults: 1,
        },
      );

      error = trade ? undefined : 'INSUFFICIENT_LIQUIDITY';
    }

    const swap =
      trade &&
      formatToSwap({
        input,
        trade,
      });

    return {
      swap,
      error,
    };
  }, [
    getUniSwapPairsData?.pairs,
    input.fromToken,
    input.toToken,
    input.fromTokenAmountTokens,
    input.toTokenAmountTokens,
  ]);
};

export default useGetSwapInfo;
