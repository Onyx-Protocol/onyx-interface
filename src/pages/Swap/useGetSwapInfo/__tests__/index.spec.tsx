import { waitFor } from '@testing-library/react';
import { Pair as PSPair, TokenAmount as PSTokenAmount } from '@uniswap/sdk';
import BigNumber from 'bignumber.js';
import React from 'react';

import fakeTokenCombinations from '__mocks__/models/tokenCombinations';
import { getUniSwapPairs } from 'clients/api';
import { UNISWAP_TOKENS } from 'constants/tokens';
import renderComponent from 'testUtils/renderComponent';

import useGetSwapInfo from '..';
import { UseGetSwapInfoInput, UseGetSwapInfoOutput } from '../types';

jest.mock('clients/api');

const fakePairs: PSPair[] = fakeTokenCombinations.map(
  ([tokenA, tokenB]) =>
    new PSPair(
      new PSTokenAmount(tokenA, new BigNumber(10).pow(tokenA.decimals).toFixed()),
      new PSTokenAmount(tokenB, new BigNumber(10).pow(tokenB.decimals).toFixed()),
    ),
);

describe('pages/Swap/useGetSwapInfo', () => {
  it('returns an error when trade consists in a wrap', async () => {
    const input: UseGetSwapInfoInput = {
      fromToken: UNISWAP_TOKENS.eth,
      toToken: UNISWAP_TOKENS.weth,
      direction: 'exactAmountIn',
    };

    let result: UseGetSwapInfoOutput | undefined;

    const TestComponent: React.FC = () => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      result = useGetSwapInfo(input);
      return <></>;
    };

    renderComponent(<TestComponent />);

    expect(result).toEqual({
      swap: undefined,
      error: 'WRAPPING_UNWRAPPING_UNSUPPORTED',
    });
  });

  it('returns an error when trade consists in an unwrap', async () => {
    const input: UseGetSwapInfoInput = {
      fromToken: UNISWAP_TOKENS.weth,
      toToken: UNISWAP_TOKENS.eth,
      direction: 'exactAmountIn',
    };

    let result: UseGetSwapInfoOutput | undefined;

    const TestComponent: React.FC = () => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      result = useGetSwapInfo(input);
      return <></>;
    };

    renderComponent(<TestComponent />);

    expect(result).toEqual({
      swap: undefined,
      error: 'WRAPPING_UNWRAPPING_UNSUPPORTED',
    });
  });

  describe('exactAmountIn', () => {
    it('returns no swap and no error if fromTokenAmountTokens is not provided', async () => {
      const input: UseGetSwapInfoInput = {
        fromToken: UNISWAP_TOKENS.usdt,
        toToken: UNISWAP_TOKENS.uni,
        toTokenAmountTokens: '1',
        direction: 'exactAmountIn',
      };

      let result: UseGetSwapInfoOutput | undefined;

      const TestComponent: React.FC = () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        result = useGetSwapInfo(input);
        return <></>;
      };

      renderComponent(<TestComponent />);

      expect(result).toEqual({
        swap: undefined,
        error: undefined,
      });
    });

    it('returns an error if no trade is found for the input provided', async () => {
      // Remove pairs containing fromToken
      const customFakePairs = fakePairs.filter(
        fakePair =>
          fakePair.token0.address !== UNISWAP_TOKENS.usdt.address &&
          fakePair.token1.address !== UNISWAP_TOKENS.usdt.address,
      );

      (getUniSwapPairs as jest.Mock).mockImplementationOnce(async () => ({
        pairs: customFakePairs,
      }));

      const input: UseGetSwapInfoInput = {
        fromToken: UNISWAP_TOKENS.usdt,
        fromTokenAmountTokens: '1',
        toToken: UNISWAP_TOKENS.uni,
        direction: 'exactAmountIn',
      };

      let result: UseGetSwapInfoOutput | undefined;

      const TestComponent: React.FC = () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        result = useGetSwapInfo(input);
        return <></>;
      };

      renderComponent(<TestComponent />);

      await waitFor(() =>
        expect(result).toEqual({
          swap: undefined,
          error: 'INSUFFICIENT_LIQUIDITY',
        }),
      );
    });

    it('returns swap in correct format if a trade is found', async () => {
      (getUniSwapPairs as jest.Mock).mockImplementationOnce(async () => ({
        pairs: fakePairs,
      }));

      const input: UseGetSwapInfoInput = {
        fromToken: UNISWAP_TOKENS.usdt,
        fromTokenAmountTokens: '1',
        toToken: UNISWAP_TOKENS.uni,
        direction: 'exactAmountIn',
      };

      let result: UseGetSwapInfoOutput | undefined;

      const TestComponent: React.FC = () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        result = useGetSwapInfo(input);
        return <></>;
      };

      renderComponent(<TestComponent />);

      await waitFor(() => expect(result?.swap).toBeDefined());
      expect(result).toMatchSnapshot();
    });
  });

  describe('exactAmountOut', () => {
    it('returns no swap and no error if toTokenAmountTokens is not provided', async () => {
      const input: UseGetSwapInfoInput = {
        fromToken: UNISWAP_TOKENS.usdt,
        fromTokenAmountTokens: '1',
        toToken: UNISWAP_TOKENS.uni,
        direction: 'exactAmountOut',
      };

      let result: UseGetSwapInfoOutput | undefined;

      const TestComponent: React.FC = () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        result = useGetSwapInfo(input);
        return <></>;
      };

      renderComponent(<TestComponent />);

      expect(result).toEqual({
        swap: undefined,
        error: undefined,
      });
    });

    it('returns an error if no trade is found for the input provided', async () => {
      (getUniSwapPairs as jest.Mock).mockImplementationOnce(async () => ({ pairs: fakePairs }));

      const input: UseGetSwapInfoInput = {
        fromToken: UNISWAP_TOKENS.usdt,
        toTokenAmountTokens: '10', // Higher amount than available liquidities in pools
        toToken: UNISWAP_TOKENS.uni,
        direction: 'exactAmountOut',
      };

      let result: UseGetSwapInfoOutput | undefined;

      const TestComponent: React.FC = () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        result = useGetSwapInfo(input);
        return <></>;
      };

      renderComponent(<TestComponent />);

      await waitFor(() =>
        expect(result).toEqual({
          swap: undefined,
          error: 'INSUFFICIENT_LIQUIDITY',
        }),
      );
    });

    it('returns swap in correct format if a trade is found', async () => {
      (getUniSwapPairs as jest.Mock).mockImplementationOnce(async () => ({
        pairs: fakePairs,
      }));

      const input: UseGetSwapInfoInput = {
        fromToken: UNISWAP_TOKENS.usdt,
        toTokenAmountTokens: '0.5',
        toToken: UNISWAP_TOKENS.uni,
        direction: 'exactAmountOut',
      };

      let result: UseGetSwapInfoOutput | undefined;

      const TestComponent: React.FC = () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        result = useGetSwapInfo(input);
        return <></>;
      };

      renderComponent(<TestComponent />);

      await waitFor(() => expect(result?.swap).toBeDefined());
      expect(result).toMatchSnapshot();
    });
  });
});
