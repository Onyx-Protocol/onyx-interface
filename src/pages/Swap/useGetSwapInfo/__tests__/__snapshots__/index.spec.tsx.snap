// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Swap/useGetSwapInfo exactAmountIn returns swap in correct format if a trade is found 1`] = `
Object {
  "error": undefined,
  "swap": Object {
    "direction": "exactAmountIn",
    "exchangeRate": "0.499248873309964947",
    "expectedToTokenAmountReceivedWei": "499248873309964947",
    "fromToken": Object {
      "address": "0xcE546eE9E8Bc0732A1e0083ef578a19F79c73D74",
      "asset": "usdt.svg",
      "decimals": 6,
      "id": "usdt",
      "symbol": "USDT",
    },
    "fromTokenAmountSoldWei": "1000000",
    "minimumToTokenAmountReceivedWei": "496765048069616862",
    "routePath": Array [
      "0xcE546eE9E8Bc0732A1e0083ef578a19F79c73D74",
      "0x651A2a9D3Fc81A5c95E8232d71a558FFd4111126",
    ],
    "toToken": Object {
      "address": "0x651A2a9D3Fc81A5c95E8232d71a558FFd4111126",
      "asset": "uni.svg",
      "decimals": 18,
      "id": "uni",
      "symbol": "UNI",
    },
  },
}
`;

exports[`pages/Swap/useGetSwapInfo exactAmountOut returns swap in correct format if a trade is found 1`] = `
Object {
  "error": undefined,
  "swap": Object {
    "direction": "exactAmountOut",
    "exchangeRate": "0.498499516455469038",
    "expectedFromTokenAmountSoldWei": "1003010",
    "fromToken": Object {
      "address": "0xcE546eE9E8Bc0732A1e0083ef578a19F79c73D74",
      "asset": "usdt.svg",
      "decimals": 6,
      "id": "usdt",
      "symbol": "USDT",
    },
    "maximumFromTokenAmountSoldWei": "1008025",
    "routePath": Array [
      "0xcE546eE9E8Bc0732A1e0083ef578a19F79c73D74",
      "0x651A2a9D3Fc81A5c95E8232d71a558FFd4111126",
    ],
    "toToken": Object {
      "address": "0x651A2a9D3Fc81A5c95E8232d71a558FFd4111126",
      "asset": "uni.svg",
      "decimals": 18,
      "id": "uni",
      "symbol": "UNI",
    },
    "toTokenAmountReceivedWei": "500000000000000000",
  },
}
`;
